asset-registry/src/lib.rs

```diff
@@ -0,0 +1,316 @@
+			T::AuthorityOrigin::ensure_origin(origin)?;
+
+			Self::do_update_asset(
+				asset_id,
+				decimals,
+				name,
+				symbol,
+				existential_deposit,
+				location,
+				additional,
+			)?;
+
+			Ok(())
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	/// Register a new asset
+	pub fn do_register_asset(
+		metadata: AssetMetadata<T::Balance, T::CustomMetadata>,
+		asset_id: Option<T::AssetId>,
+	) -> DispatchResult {
+		let (asset_id, metadata) = T::AssetProcessor::pre_register(asset_id, metadata)?;
+
+		Self::do_register_asset_without_asset_processor(metadata.clone(), asset_id.clone())?;
+
+		T::AssetProcessor::post_register(asset_id, metadata)?;
+
+		Ok(())
+	}
+
+	/// Like do_register_asset, but without calling pre_register and
+	/// post_register hooks.
+	/// This function is useful in tests but it might also come in useful to
+	/// users.
+	pub fn do_register_asset_without_asset_processor(
+		metadata: AssetMetadata<T::Balance, T::CustomMetadata>,
+		asset_id: T::AssetId,
+	) -> DispatchResult {
+		Metadata::<T>::try_mutate(&asset_id, |maybe_metadata| -> DispatchResult {
+			// make sure this asset id has not been registered yet
+			ensure!(maybe_metadata.is_none(), Error::<T>::ConflictingAssetId);
+
+			*maybe_metadata = Some(metadata.clone());
+
+			if let Some(ref location) = metadata.location {
+				Self::do_insert_location(asset_id.clone(), location.clone())?;
+			}
+
+			Ok(())
+		})?;
+
+		Self::deposit_event(Event::<T>::RegisteredAsset { asset_id, metadata });
+
+		Ok(())
+	}
+
+	pub fn do_update_asset(
+		asset_id: T::AssetId,
+		decimals: Option<u32>,
+		name: Option<Vec<u8>>,
+		symbol: Option<Vec<u8>>,
+		existential_deposit: Option<T::Balance>,
+		location: Option<Option<VersionedMultiLocation>>,
+		additional: Option<T::CustomMetadata>,
+	) -> DispatchResult {
+		Metadata::<T>::try_mutate(&asset_id, |maybe_metadata| -> DispatchResult {
+			match maybe_metadata {
+				None => Err(Error::<T>::AssetNotFound.into()),
+				Some(ref mut metadata) => {
+					if let Some(decimals) = decimals {
+						(*metadata).decimals = decimals;
+					}
+
+					if let Some(name) = name {
+						(*metadata).name = name;
+					}
+
+					if let Some(symbol) = symbol {
+						(*metadata).symbol = symbol;
+					}
+
+					if let Some(existential_deposit) = existential_deposit {
+						(*metadata).existential_deposit = existential_deposit;
+					}
+
+					if let Some(location) = location {
+						// Update LocationToAssetId only if location changed
+						if location != metadata.location {
+							// remove the old location lookup if it exists
+							if let Some(ref location) = metadata.location {
+								let location: MultiLocation =
+									location.clone().try_into().map_err(|()| Error::<T>::BadVersion)?;
+								LocationToAssetId::<T>::remove(location);
+							}
+
+							// insert new location
+							if let Some(ref new_location) = location {
+								Self::do_insert_location(asset_id.clone(), new_location.clone())?;
+							}
+						}
```

Comment:
would be better to use `LocationToAssetId::try_mutate(location, |maybe_asset_id| {....})`

Reply:
We are dealing with 2 different keys though - we have `old_location` and `new_location`, so it can't be combined into a single `try_mutate`. Setting `new_location` does use `try_mutate` inside `do_insert_location`, but for `old_location` the remove is more straightforward